caas/containers.go

```diff
@@ -3,67 +3,77 @@
-}
+import "github.com/juju/errors"
 
 // FileSet defines a set of files to mount
 // into the container.
 type FileSet struct {
-	Name      string            `yaml:"name"`
-	MountPath string            `yaml:"mount-path"`
-	Files     map[string]string `yaml:"files"`
+	Name      string            `yaml:"name" json:"name"`
+	MountPath string            `yaml:"mountPath" json:"mountPath"`
+	Files     map[string]string `yaml:"files" json:"files"`
+}
+
+// ContainerPort defines a port on a container.
+type ContainerPort struct {
+	Name          string `yaml:"name,omitempty" json:"name,omitempty"`
+	ContainerPort int32  `yaml:"containerPort" json:"containerPort"`
+	Protocol      string `yaml:"protocol" json:"protocol"`
+}
+
+// ProviderContainer defines a provider specific container.
+type ProviderContainer interface {
+	Validate() error
 }
 
 // ContainerSpec defines the data values used to configure
 // a container on the CAAS substrate.
 type ContainerSpec struct {
-	Name      string            `yaml:"name"`
-	ImageName string            `yaml:"image-name"`
-	Ports     []ContainerPort   `yaml:"ports,omitempty"`
-	Config    map[string]string `yaml:"config,omitempty"`
-	Files     []FileSet         `yaml:"files,omitempty"`
+	Name  string          `yaml:"name"`
+	Image string          `yaml:"image,omitempty"`
+	Ports []ContainerPort `yaml:"ports,omitempty"`
+
+	Config map[string]string `yaml:"config,omitempty"`
+	Files  []FileSet         `yaml:"files,omitempty"`
+
+	// ProviderContainer defines config which is specific to a substrate, eg k8s
+	ProviderContainer `yaml:"-"`
 }
 
 // PodSpec defines the data values used to configure
 // a pod on the CAAS substrate.
 type PodSpec struct {
-	Containers          []ContainerSpec `yaml:"containers"`
-	OmitServiceFrontend bool            `yaml:"omit-service-frontend"`
+	Containers          []ContainerSpec `yaml:"-"`
+	OmitServiceFrontend bool            `yaml:"omitServiceFrontend"`
+}
+
+// Validate returns an error if the spec is not valid.
+func (spec *PodSpec) Validate() error {
+	for _, c := range spec.Containers {
+		if err := c.Validate(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
 }
 
-// ParsePodSpec parses a YAML string into a PodSpec struct.
-func ParsePodSpec(in string) (*PodSpec, error) {
-	var spec PodSpec
-	if err := yaml.Unmarshal([]byte(in), &spec); err != nil {
-		return nil, errors.Trace(err)
+// Validate is defined on ProviderContainer.
+func (spec *ContainerSpec) Validate() error {
+	if spec.Name == "" {
+		return errors.New("spec name is missing")
 	}
-	if len(spec.Containers) == 0 {
-		return nil, errors.New("require at least one pod spec")
+	if spec.Image == "" {
+		return errors.New("spec image is missing")
 	}
-	for _, container := range spec.Containers {
-		if container.Name == "" {
-			return nil, errors.New("spec name is missing")
+	for _, fs := range spec.Files {
+		if fs.Name == "" {
+			return errors.New("file set name is missing")
```

Comment:
This is iterating over a few files and if one name is missing, we will bail? Did you intend that or did you actually want to log and 'continue'? 
The same goes for mount path check below..

Reply:
Intended to bail because even one missing name messes up the entire config